FEXCore/Source/Utils/ArchHelpers/Arm64.cpp

```diff
@@ -1830,178 +1830,237 @@ HandleUnalignedAccess(FEXCore::Core::InternalThreadState* Thread, UnalignedHandl
+      return std::make_pair(true, 4);
+    } else {
+      LogMan::Msg::EFmt("Unhandled JIT SIGBUS CASAL: PC: 0x{:x} Instruction: 0x{:08x}\n", ProgramCounter, PC[0]);
+      return NotHandled;
+    }
+  } else if ((Instr & ArchHelpers::Arm64::ATOMIC_MEM_MASK) == ArchHelpers::Arm64::ATOMIC_MEM_INST) { // Atomic memory op
+    if (ArchHelpers::Arm64::HandleAtomicMemOp(Instr, GPRs)) {
+      // Skip this instruction now
+      return std::make_pair(true, 4);
+    } else {
+      uint8_t Op = (PC[0] >> 12) & 0xF;
+      LogMan::Msg::EFmt("Unhandled JIT SIGBUS Atomic mem op 0x{:02x}: PC: 0x{:x} Instruction: 0x{:08x}\n", Op, ProgramCounter, PC[0]);
+      return NotHandled;
+    }
+  } else if ((Instr & ArchHelpers::Arm64::LDAXR_MASK) == ArchHelpers::Arm64::LDAXR_INST) { // LDAXR*
+    uint64_t BytesToSkip = ArchHelpers::Arm64::HandleAtomicLoadstoreExclusive(ProgramCounter, GPRs);
+    if (BytesToSkip) {
+      // Skip this instruction now
+      return std::make_pair(true, BytesToSkip);
+    } else {
+      LogMan::Msg::EFmt("Unhandled JIT SIGBUS LDAXR: PC: 0x{:x} Instruction: 0x{:08x}\n", ProgramCounter, PC[0]);
+      return NotHandled;
+    }
+  } else if ((Instr & ArchHelpers::Arm64::LDAXP_MASK) == ArchHelpers::Arm64::LDAXP_INST) { // LDAXP
+    // Should be compare and swap pair only. LDAXP not used elsewhere
+    uint64_t BytesToSkip = ArchHelpers::Arm64::HandleCASPAL_ARMv8(Instr, ProgramCounter, GPRs);
+    if (BytesToSkip) {
+      // Skip this instruction now
+      return std::make_pair(true, BytesToSkip);
+    }
   }
+
+  // Lock code mutex during any SIGBUS handling that potentially changes code.
+  // Need to be careful to not read any code part-way through modification.
+  FEXCore::Utils::SpinWaitLock::UniqueSpinMutex lk(&InlineTail->SpinLockFutex);
+
+  constexpr uint32_t LDRBase = 0b0011'1000'0111'1111'0110'1000'0000'0000;
+  constexpr uint32_t STRBase = 0b0011'1000'0011'1111'0110'1000'0000'0000;
+  constexpr uint32_t LDSTRegisterMask = 0b0011'1011'0010'0000'0000'1100'0000'0000;
+
+  constexpr uint32_t LDURBase = 0b0011'1000'0100'0000'0000'0000'0000'0000;
+  constexpr uint32_t STURBase = 0b0011'1000'0000'0000'0000'0000'0000'0000;
+  constexpr uint32_t LDSTUnscaledMask = 0b0011'1011'0010'0000'0000'1100'0000'0000;
+
+  constexpr uint32_t STPBase = 0b0010'1001'0000'0000'0000'0000'0000'0000;
+  constexpr uint32_t LDSTPMask = 0b0011'1011'1000'0000'0000'0000'0000'0000;
+
+  if ((Instr & LDAXR_MASK) == LDAR_INST ||  // LDAR*
+      (Instr & LDAXR_MASK) == LDAPR_INST) { // LDAPR*
+    uint32_t LDR = LDRBase;
+    LDR |= Size << 30;
+    LDR |= AddrReg << 5;
+    LDR |= DataReg;
+    if (HandleType != UnalignedHandlerType::NonAtomic) {
+      ///< Ordering matters with cross-thread visibility!
+      std::atomic_ref<uint32_t>(PC[1]).store(DMB_LD, std::memory_order_relaxed); // Back-patch the half-barrier.
+    }
+    std::atomic_ref<uint32_t>(PC[0]).store(LDR, std::memory_order_relaxed);
+    ClearICache(&PC[0], 16);
```

Comment:
Updated to pass in the number of bytes modified instead of hardcoded to 16. Since it's less than a cacheline the icache clear will still only do a single iteration of a loop so behaviour doesn't change.